### ISSUE TITLE
How can i add metadata to the resource?

### ISSUE THREAD

**Issue by humdafadumda** [joukevandermaas/saule#78]
i want to add some info about total records, how can i add this info?
(Using EmberJS) -> meta
example, 
{
  "post": {
    "id": 1,
    "title": "Progressive Enhancement is Dead",
    "comments": ["1", "2"],
    "links": {
      "user": "/people/tomdale"
    },
    // ...
  },

  "meta": {
    "total": 100
  }
}

**Comment by joukevandermaas** [comment]
Saule does not currently support meta data. I hoped someone would make the request so I could figure out what is required of this feature. I hope you don't mind me asking you some questions to sketch out the feature.

How many action methods are going to add this `total` value? Will it be e.g. all methods that return `IEnumerable` or only this specific one? Are there any other values you'd like to add?

**Comment by humdafadumda** [comment]
According to the JSON API spec, a document MUST contain at least one of the following top-level members:
data: the document's "primary data"
errors: an array of error objects (cannot coexist with data)
meta: a meta object that contains non-standard meta-information.
--> Meta Information
Where specified, a meta member can be used to include non-standard meta-information. The value of each meta member MUST be an object (a "meta object").
Any members MAY be specified within meta objects.

So i am thinking more of a"marker" interface IMetaData that can be serialized in the ApiResource, i think it would be enough only for server - client communication

**Comment by markwalsh-liverpool** [comment]
:+1: the spec dictates that meta is compulsory

**Comment by joukevandermaas** [comment]
@markwalsh-liverpool no it does not. [It dictates](http://jsonapi.org/format/#document-top-level) that a document must contain **at least one** of the following: `data`, `errors`, `meta`. Saule currently always generates either `data` or `errors`, so it complies with this section of the spec.

That said, I'm all for adding `meta` support, and I'll likely do so in the near(ish) future.

**Comment by markwalsh-liverpool** [comment]
Sorry my bad, it does!

**Comment by joe-zhang3** [comment]
Our project is using Saule and it is pretty cool. Thank you.

But we also facing the "meta" problem as we want to include a "total-records" when user performs a search action. JSON API suggests to use meta to do this, (http://jsonapi.org/examples/#pagination), and we agree with that. 

Since this issue was removed from the 1.5 milestone, I wonder is there any other plan to address this?

Appreciate.

**Comment by joukevandermaas** [comment]
@joe-zhang3 I did not have time to look at this for the 1.5 and 1.6 releases. I do still want this, but I am unsure about the design. I have a few thoughts but I don't really like any of them. Maybe you can elaborate on some of your use-cases or thoughts on how you would like this to work.

**Comment by joe-zhang3** [comment]
@joukevandermaas , this is the response that we actually want:
```json
{
  "meta": {
    "total-pages": 13
  },
  "data": [
    {
      "type": "articles",
      "id": "3",
      "attributes": {
        "title": "JSON API paints my bikeshed!",
        "body": "The shortest article. Ever.",
        "created": "2015-05-22T14:56:29.000Z",
        "updated": "2015-05-22T14:56:28.000Z"
      }
    }
  ],
  "links": {
    "self": "http://example.com/articles?page[number]=3&page[size]=1",
    "first": "http://example.com/articles?page[number]=1&page[size]=1",
    "prev": "http://example.com/articles?page[number]=2&page[size]=1",
    "next": "http://example.com/articles?page[number]=4&page[size]=1",
    "last": "http://example.com/articles?page[number]=13&page[size]=1"
  }
}
```
From our point of view, "meta" acts very similar as "attributes". Like we can use this in our code:
```csharp
Meta("TotalRecords");
```
And when the response is rendering, it will try to find the property with the same name and handle it like how attribute is being handled.

**Comment by joukevandermaas** [comment]
The attributes are on the individual records though, while the meta hash is about the whole response. Maybe a virtual method `GetMetadataForResponse` on `ApiResource`:

```csharp
protected virtual object GetMetadataForResponse(object response, Type resourceType)
{
    return null;
}
```

You could override it as follows:

```csharp
protected override object GetMetadataForResponse(object response, Type resourceType)
{
    if (resourceType != typeof(Article))
    {
        return null;
    }
    return ((MyCollection<Article>)response).TotalRecords
}
```

Would that serve your needs?

**Comment by joe-zhang3** [comment]
Yes, it does for this scenario, thanks. 

However, "meta" is not only about the whole response. Based on the specification of JSON API 

`a meta object containing non-standard meta-information about a resource that can not be represented as an attribute or relationship`

meta can also be part of resource object, which means it might shows up in the "data" as well. 
(and it can be part of a "relationship" and can be part of the link object etc).

But we don't need that so far as long as the top level meta works. 

thanks again.

**Comment by joukevandermaas** [comment]
I implemented this in #158. The latest pre-release should have this feature.
